### PR TITLE
Remove uses of `fail` and `MonadFail`

### PR DESCRIPTION
**Pull request by brianhuffman**
Uses of `fail` in saw-core mostly indicate not-well-thought-out error handling. Many uses of `fail` are replaced with calls to `panic`; other uses of `fail` are replaced with thrown exceptions.

Hopefully changes in this PR will make it much easier to adapt the code for GHC 8.8.

### CODE REVIEW

File: src/Verifier/SAW/Conversion.hs [hunk 1]
```diff
@@ -394,7 +394,7 @@ mkTupleSelector :: Int -> Term -> TermBuilder Term
 mkTupleSelector i t
   | i == 1 = mkTermF (FTermF (PairLeft t))
   | i > 1  = mkTermF (FTermF (PairRight t)) >>= mkTupleSelector (i - 1)
-  | otherwise = fail "mkTupleSelector: non-positive index"
+  | otherwise = error "mkTupleSelector: non-positive index"
```
Comment (robdockins) [inline src/Verifier/SAW/Conversion.hs]: Why error instead of panic?

Comment (brianhuffman) [reply src/Verifier/SAW/Conversion.hs]: That was a mistake, it should be `panic`. Actually I was thinking that we could get by without any error case at all in that function, by changing the argument type from `Int` to `Natural`, and making the tuple indexing 0-based to match Cryptol, but I think that's out of scope for this PR. I'll just change it to `panic`.

**Comment by brianhuffman** [comment]
I tried compiling with ghc 8.8.1, and there is at least one more use of `fail` that we need to address: in `Grammar.y`, the function `parseError` calls `fail` in the `Parser` monad, but the `Parser` monad doesn't have any way to model failure, and is not an instance of `MonadFail`.

**Comment by robdockins** [comment]
See 1ef07681d3b94085e7efdbf66baaae043f32f312 for a potential solution to the `parseError` issue.

Alternately, we could make the `Parse` monad use `ExceptT` or somesuch.

File: src/Verifier/SAW/Grammar.y [hunk 2]
```diff
@@ -246,17 +247,17 @@ data ParserState = PS { psInput :: AlexInput
                       , psErrors :: [PosPair ParseError]
```
Comment (robdockins) [inline src/Verifier/SAW/Grammar.y]: I don't think we need this `ParserState` datatype anymore, or the `ErrorList` type.

Comment (brianhuffman) [reply src/Verifier/SAW/Grammar.y]: I removed those, and some other dead code, in a07d0a8.

**Pull request merged**